File: app/src/main/java/com/example/clearentwrapperdemo/feature/PairingFragment.java
package com.example.clearentwrapperdemo.feature;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.clearent.idtech.android.wrapper.model.ReaderStatus;
import com.example.clearentwrapperdemo.databinding.FragmentPairingBinding;

import java.util.List;
import java.util.function.Consumer;

public class PairingFragment extends Fragment implements ReadersListAdapter.ItemClickListener {

    private FragmentPairingBinding binding;

    private final List<ReaderStatus> readers;
    private final Consumer<ReaderStatus> selectReader;
    private final Runnable cancel;

    public PairingFragment(List<ReaderStatus> list, Consumer<ReaderStatus> selectReader, Runnable cancel) {
        this.readers = list;
        this.selectReader = selectReader;
        this.cancel = cancel;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentPairingBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setupViews();
    }

    private void setupViews() {
        ReadersListAdapter adapter = new ReadersListAdapter(readers, this);
        binding.readersList.setLayoutManager(new LinearLayoutManager(requireContext()));
        binding.readersList.setAdapter(adapter);
        if (readers.isEmpty())
            binding.pairingNoReadersInfo.setVisibility(View.VISIBLE);
        else
            binding.pairingNoReadersInfo.setVisibility(View.GONE);
        binding.pairingCancelButton.setOnClickListener(view -> cancel.run());
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        binding = null;
    }

    @Override
    public void onItemClick(ReaderStatus readerStatus) {
        selectReader.accept(readerStatus);
    }
}

File: app/src/main/java/com/example/clearentwrapperdemo/feature/PairingFragment.kt
package com.example.clearentwrapperdemo.feature

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.clearent.idtech.android.wrapper.model.ReaderStatus
import com.example.clearentwrapperdemo.data.DataStatus
import com.example.clearentwrapperdemo.databinding.FragmentPairingBinding

class PairingFragment(
    private val state: DataStatus.ReadersList,
    private val selectReader: (ReaderStatus) -> Unit,
    private val cancel: () -> Unit
) : Fragment(), ReadersListAdapter.OnItemClickListener {

    private var _binding: FragmentPairingBinding? = null
    private val binding get() = _binding!!

    private val readersListAdapter = ReadersListAdapter(this)

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentPairingBinding.inflate(inflater, container, false)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setupViews()
    }

    private fun setupViews() {
        binding.apply {
            readersList.apply {
                adapter = readersListAdapter
                layoutManager = LinearLayoutManager(requireContext())
                readersListAdapter.submitList(state.list)

                if (state.list.isEmpty())
                    pairingNoReadersInfo.visibility = View.VISIBLE
                else
                    pairingNoReadersInfo.visibility = View.GONE

                pairingCancelButton.setOnClickListener {
                    cancel()
                }
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onItemClick(reader: ReaderStatus) {
        selectReader(reader)
    }
}